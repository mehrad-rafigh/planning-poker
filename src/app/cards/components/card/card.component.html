<div
  class="card"
  [ngStyle]="{ 'background-color': cardColor.backgroundColor }"
  [ngClass]="setFontColor()"
  (click)="handleClick()"
>
  <span *ngIf="!detailMode" class="top-right quarter-h">{{ value }}</span>
  <span *ngIf="detailMode" class="half-h" [ngClass]="setDetailMode()">{{
    value
  }}</span>
  <span *ngIf="!detailMode" class="bottom-left quarter-h">{{ value }}</span>
</div>
