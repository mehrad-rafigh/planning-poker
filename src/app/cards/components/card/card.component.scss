@import "../../../../mixins";

span {
  font-weight: bold;
}

.card {
  display: flex;
  flex-flow: column nowrap;
  box-sizing: border-box;
  height: 200px;
  width: auto;
  border-color: lightslategray;
  position: relative;
  box-shadow: 0 4px 8px 0 rgba(0, 0, 0, 0.2);
  transition: 0.3s;
  border-radius: 15px;

  @include mobile() {
    height: 250px;
    font-size: 26px;
  }

  @include tablet() {
    height: 300px;
    font-size: 24px;
  }

  @include above-tablet() {
    height: 350px;
    font-size: 22px;
  }
}

.top-right {
  text-align: right;
  margin-top: 32px;
  margin-right: 32px;
}

.bottom-left {
  margin-bottom: 32px;
  margin-left: 32px;
  position: absolute;
  bottom: 0;
  height: auto !important;
}

.quarter-h {
  height: 25%;
}

.half-h {
  height: 50%;
}

.detail-mode {
  height: unset;
  margin: auto;
  font-size: 128px;
}

.font-white {
  color: white;
}
