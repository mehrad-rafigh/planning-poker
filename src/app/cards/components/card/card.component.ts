import {
  Component,
  OnInit,
  ChangeDetectionStrategy,
  Input,
  EventEmitter,
  Output
} from "@angular/core";

@Component({
  selector: "app-card",
  templateUrl: "./card.component.html",
  styleUrls: ["./card.component.scss"],
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class CardComponent implements OnInit {
  @Input() value: string;
  @Input() cardColor: { backgroundColor: string; fontColor: string };
  @Input() detailMode = false;
  @Output() clickEvent = new EventEmitter<any>();
  constructor() {}

  ngOnInit(): void {}

  handleClick(): void {
    this.clickEvent.emit({
      value: this.value,
      card: {
        backgroundColor: this.cardColor.backgroundColor,
        fontColor: this.cardColor.fontColor
      }
    });
  }

  setDetailMode(): string {
    if (this.detailMode) {
      return "detail-mode";
    }
  }

  setFontColor(): string {
    if (this.cardColor.fontColor === "white") {
      return "font-white";
    }
  }
}
