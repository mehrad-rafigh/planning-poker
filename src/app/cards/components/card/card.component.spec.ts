import { async, ComponentFixture, TestBed } from "@angular/core/testing";

import { CardComponent } from "./card.component";
import { ChangeDetectionStrategy, DebugElement } from "@angular/core";
import { By } from "@angular/platform-browser";

describe("CardComponent", () => {
  let component: CardComponent;
  let fixture: ComponentFixture<CardComponent>;
  let debugElement: DebugElement;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [CardComponent]
    })
      .overrideComponent(CardComponent, {
        set: {
          changeDetection: ChangeDetectionStrategy.Default
        }
      })
      .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(CardComponent);
    component = fixture.componentInstance;
    debugElement = fixture.debugElement;

    component.value = "0.5";
    component.cardColor = {
      fontColor: "white",
      backgroundColor: "rgb(100,100,100)"
    };
    component.detailMode = false;
    fixture.detectChanges();
  });

  it("should create", () => {
    expect(component).toBeTruthy();
  });

  it("should match the snapshot", () => {
    expect(fixture).toMatchSnapshot();
  });

  it("should render value in top right and bottom left corners", () => {
    const spanTopRightCorner: HTMLSpanElement = debugElement.query(
      By.css(".top-right")
    ).nativeElement;
    const spanBottomLeftCorner: HTMLSpanElement = debugElement.query(
      By.css(".bottom-left")
    ).nativeElement;

    expect(spanBottomLeftCorner.innerHTML).toBe("0.5");
    expect(spanTopRightCorner.innerHTML).toBe("0.5");
  });

  it("should render value in center when detail mode is on", () => {
    component.detailMode = true;
    fixture.detectChanges();

    const spanInCenterDebugEl = debugElement.query(By.css(".half-h"));
    const spanTopRightCorner = debugElement.query(By.css(".top-right"));
    const spanBottomLeftCorner = debugElement.query(By.css(".bottom-left"));

    expect(spanInCenterDebugEl.classes["detail-mode"]).toBeTruthy();
    expect(spanInCenterDebugEl.nativeElement.innerHTML).toBe("0.5");
    expect(spanTopRightCorner).toBeFalsy();
    expect(spanBottomLeftCorner).toBeFalsy();
    expect(fixture).toMatchSnapshot();
  });

  it("should handle click events", () => {
    component.clickEvent.emit = jest.fn();

    const divElementDebugElement = debugElement.query(By.css(".card"));
    divElementDebugElement.triggerEventHandler("click", null);

    expect(component.clickEvent.emit).toHaveBeenCalledWith({
      value: "0.5",
      card: {
        backgroundColor: "rgb(100,100,100)",
        fontColor: "white"
      }
    });
  });
});
