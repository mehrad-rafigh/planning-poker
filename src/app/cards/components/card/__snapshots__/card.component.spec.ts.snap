// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardComponent should match the snapshot 1`] = `
<app-card
  cardColor={[Function Object]}
  clickEvent={[Function EventEmitter]}
  detailMode="false"
  value={[Function String]}
>
  <div
    class="card font-white"
    style="background-color: rgb(100, 100, 100);"
  >
    
    <span
      class="top-right quarter-h"
    >
      0.5
    </span>
    
    
    <span
      class="bottom-left quarter-h"
    >
      0.5
    </span>
  </div>
</app-card>
`;

exports[`CardComponent should render value in center when detail mode is on 1`] = `
<app-card
  cardColor={[Function Object]}
  clickEvent={[Function EventEmitter]}
  detailMode={[Function Boolean]}
  value={[Function String]}
>
  <div
    class="card font-white"
    style="background-color: rgb(100, 100, 100);"
  >
    
    
    <span
      class="half-h detail-mode"
    >
      0.5
    </span>
    
  </div>
</app-card>
`;
